[windows] Add ability to tag IIS services by site name in addition to…

derekwbrown opened the pull request:
… site id.

### What does this PR do?

### Additional Notes

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes
Install windows with IIS.
install agent
Generate traffic to IIS.
hit the debugging endpoint:
`iwr -UseBasicParsing -DisableKeepAlive http://localhost:3333/network_tracer/debug/http_monitoring | convertfrom-json`

You should see output indicating that the site id and site name are now correct, e.g.

`DynamicTags : {http.iis.app_pool:DefaultAppPool, http.iis.site:1, http.iis.sitename:Default Web Site,
              service:DefaultAppPool}`


### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

derekwbrown committed:
[windows] Add ability to tag IIS services by site name in addition to site id.

update iis xml parsing

Add unit test
Fixes stop problem

Update so that the tag values (site names, app pools) don't have spaces

update unit test to reflect removing spaces from site names

clarkb7 commented on pkg/network/etw/etw-http-service.go:
```diff
@@ -746,8 +751,9 @@ func httpCallbackOnHTTPRequestTraceTaskDeliver(eventInfo *C.DD_ETW_EVENT_INFO) {
 	}
 
-	httpConnLink.http.AppPool = appPool
+	httpConnLink.http.AppPool = strings.Replace(appPool, " ", "_", -1)
```
why the change in how the `appPool` is stored?

derekwbrown replied:
because it becomes a tag and it can't have spaces.

clarkb7 replied:
can you rename the `http.AppPool` and `http.SiteName` attributes as well, to indicate they contain tag names and not the actual application pool and site name.

iglendd replied:
I actually do not like renaming functions or vars things here (in this module). They are converted to tags at different layer. Perhaps that where it should be replace spaces with underscores.

clarkb7 commented on pkg/util/winutil/iisconfig.go:
```diff
@@ -0,0 +1,171 @@
+		if err != nil {
+			return err
+		}
+		idmap[uint32(id)] = strings.Replace(site.Name, " ", "_", -1)
```
why not store the site name as is?

derekwbrown replied:
because the site name becomes a tag, and tags shouldn't hvae spaces.  And it's much more efficient to do it once when we parse it than to to the string convert every time we access it.

clarkb7 replied:
Are the `SiteName` or `AppPool` variables used anywhere else that might expect the real value instead of a tag value?

If you must store them as tag values the variable name should reflect that (e.g. `SiteName_tag`)

derekwbrown replied:
it is only used as a tag.
I can change the variable name.
The reason I don't want to make the transformation when the tag is made is that is potentially done thousands of times per second, vs. doing once when we read.

derekwbrown committed:
Result of go mod tidy

rename function and result map as suggested in review

Rename a member.
Also fix (unrelated) typo in resource

rename more variables....

Revert "rename more variables...."

This reverts commit 2949b56eace71bec99379f222cba0cae22c5b2c3.

Revert "Rename a member."

This reverts commit f09b297fd96af077a1abe5ab7c0e65c36eac78d4.

Revert "rename function and result map as suggested in review"

This reverts commit fdbbe17125eb926f2398270e1c34ee7b7248cdc3.

Remove unnecessary conversion of space to _

richl-dd commented on pkg/util/winutil/iisconfig.go:
```diff
@@ -0,0 +1,171 @@
+
+var (
+	// make global so that we can override for tests.
+	iisCfgPath = filepath.Join(os.Getenv("windir"), "System32", "inetsrv", "config", "applicationHost.config")
```
Should we make this configurable in case a customer installed in a non-default location?

richl-dd replied:
👍

richl-dd commented on pkg/util/winutil/iisconfig.go:
```diff
@@ -0,0 +1,171 @@
+
+	// check for existence
+	_, err = os.Stat(iisCfgPath)
+	if os.IsNotExist(err) {
```
Can we log an error here so it shows up in a flare and we can see that the path didn't exist at the expected location. Will help support cases since we know that is the reason.

derekwbrown replied:
since the error is returned, it is already logged by the caller.

richl-dd reviewed (commented):
Aside from comment on configurable IIS config path, looks good to me.

clarkb7 commented on pkg/network/etw/etw-http-service.go:
```diff
@@ -152,15 +153,16 @@ type Http struct {
 	// Url           string
-	SiteID uint32
+	SiteID   uint32
+	SiteName string
```
Can you rename this member `SiteNameAsTag` as well

clarkb7 reviewed (approved):
it might be worth it to add a comment to the struct explaining your reasoning for storing the tag names instead of the real names from the previous comment so someone doesn't refactor it later
> The reason I don't want to make the transformation when the tag is made is that is potentially done thousands of times per second, vs. doing once when we read.
